### ISSUE TITLE
Support nonterminals participating in P&A.

### ISSUE THREAD

**Issue by teo-tsirpanis** [teo-tsirpanis/Farkle#41]
Consider this grammar represented in Bison syntax:

```
%left '|' "or"
%left '&' "and"
%precedence '!' "not"

%start expr

or_op: '|' | "or";
and_op: '&' | "and";

expr:
  expr or_op expr
  | expr and_op expr
  | "x";
```

Farkle fails to build this grammar because the `expr <op> expr` productions don't have any P&A[^PnA], and that's because the operator terminals are indirectly specified. The current workaround is to inline `or_op` and `and_op`. There are two ways to solve this:

*
    __Recognize nonterminals in associativity groups.__ The constructors of associativity groups accept an array of objects, which can be `DesigntimeFarkle`s for terminals, strings for literals, or arbitrary objects for contextual precedence tokens, ignoring everything else. We should also recognize `DesigntimeFarkle`s for nonterminals.

    With this change the P&A of productions without contextual precedence would become the P&A of the last terminal __or nonterminal__ of the production that has P&A. This would be the equivalent of writing `%left <op>`.
*
    __Infer the P&A of nonterminals whose all productions have the same P&A.__ This would make the above grammar work with no modifications to the P&A rules. If we go down this way we have to watch out for recursive nonterminals.

Judging from Bison's documentation, it doesn't do the things proposed above; I wonder why. Should we even do such comparisons? In any case I don't see any downsides; it won't affect those that don't use them.

[^PnA]: Precedence & Associativity

**Comment by teo-tsirpanis** [comment]
Update: item 2 is not as simple as it sounds when considering more complex cases like recursive nonterminals. Improving Farkle's conflict resolution mechanism is an interesting area to explore, but it must be done carefully.

**Comment by teo-tsirpanis** [comment]
Turns out item 1 can't be done either. Conflict resolution deals with conflicts between shifting _terminals_ and reducing productions, and trying to determine the nonterminal we will GOTO after shifting is impossible in all non-trivial cases (and likely in all cases).

With that in mind, if we included nonterminals in determining the P&A of a production, the conflicts in the OP's grammar would be resolved by specifying both `<op>` and its terminals, which defeats the purpose. The correct workaround is to either inline `or_op` and `and_op`, or make them terminals and specify the alternation in the regex. Closing as not planned.
